---
title: "Figure3"
author: "Jiang junyao"

output: 
  prettydoc::html_pretty:
    toc: true
    theme: cayman
    highlight: github
---
Here is the script to generate figures in Figure 3 and related supplementary figures in the PolyATAC manuscript.

## Load data
```{r, message=FALSE,warning=FALSE}
library(tidyverse)
library(ggpubr)
library(pheatmap)
library(Hmisc)
library(corrplot)
library(Seurat)
library(BuenColors)
source('/data/jiangjunyao/polyATAC/github_script/help_function/figure3_help_function.R')
pm_smooth = readRDS('/data/jiangjunyao/polyATAC/github_script/data/pm_smooth.rds')
sp_smooth = readRDS('/data/jiangjunyao/polyATAC/github_script/data/sp_smooth.rds')
obj = readRDS('/data/jiangjunyao/polyATAC/nmp_final/nmp_sp_pm_only_rna.rds')
group_df_20241115 = read.csv('/data/jiangjunyao/polyATAC/nmp_final/nmp_pseudotime_module/only_rna/group_df_20241115.csv',
                             row.names = 1)
Color1 <- c(rgb(72 / 255, 85 / 255, 167 / 255), rgb(255 / 255, 255 / 255, 255 / 255),
            rgb(239 / 255, 58 / 255, 37 / 255))
fate_col = c('#4DB6AC',
             '#0D47A1','#1B5E20','#9ECAE1','#C5E1A5','#D3D3D3')
names(fate_col) = c('Balance','Paraxial mesoderm','Spinal cord',"Paraxial_mesoderm_bias","Spinal_cord_bias",'other')
ct_col = c('#0D47A1','#1B5E20','#4DB6AC')
names(ct_col) = c('Paraxial mesoderm','Spinal cord','Neuromesodermal progenitors')

```

## Plot NMP differentiation trajectory

```{r, message=FALSE,warning=FALSE}
plot_df = obj@meta.data
plot_df$DiffusionMap_1 = obj@reductions$diffmap@cell.embeddings[,1]
plot_df$DiffusionMap_2 = obj@reductions$diffmap@cell.embeddings[,2]
ggplot(plot_df,aes(x=DiffusionMap_1,y=DiffusionMap_2,color=celltype))+
  geom_point(size=1.2)+scale_color_manual(values = fate_col)+
  theme_void()+theme(text = element_text(size=16))+theme(legend.position = "none")
ggsave('/data/jiangjunyao/polyATAC/polyatac_figure/all_embedding_no_legend/fig3/NMP diffusion map embedding celltype.pdf',
       width = 10,height = 8)
ggplot(plot_df,aes(x=DiffusionMap_1,y=DiffusionMap_2,color=celltype_ori))+
  geom_point(size=1.2)+scale_color_manual(values = ct_col)+
  theme_void()+theme(text = element_text(size=16))+theme(legend.position = "none")
ggsave('/data/jiangjunyao/polyATAC/polyatac_figure/all_embedding_no_legend/fig3/NMP diffusion map embedding celltype_original.pdf',
       width = 10,height = 8)
ggplot(plot_df,aes(x=DiffusionMap_1,y=DiffusionMap_2,color=Pseudotime))+
  geom_point(size=1.2)+scale_color_gradientn(colors = rev(jdb_palette("brewer_marine")))+
  theme_void()+theme(text = element_text(size=16))+theme(legend.position = "none")
ggsave('/data/jiangjunyao/polyATAC/polyatac_figure/all_embedding_no_legend/fig3/NMP diffusion map embedding pseudotime.pdf',
       width = 10,height = 8)
ggplot(plot_df,aes(x=DiffusionMap_1,y=DiffusionMap_2,color=sp_score))+
  geom_point(size=1.2)+scale_color_viridis_c()+
  theme_void()+theme(text = element_text(size=16))+theme(legend.position = "none")
ggsave('/data/jiangjunyao/polyATAC/polyatac_figure/all_embedding_no_legend/fig3/NMP diffusion map embedding spinal cord bias score.pdf',
       width = 10,height = 8)
ggplot(plot_df,aes(x=DiffusionMap_1,y=DiffusionMap_2,color=pm_score))+
  geom_point(size=1.2)+scale_color_viridis_c()+
  theme_void()+theme(text = element_text(size=16))+theme(legend.position = "none")
ggsave('/data/jiangjunyao/polyATAC/polyatac_figure/all_embedding_no_legend/fig3/NMP diffusion map embedding paraxial mesoderm bias score.pdf',
       width = 10,height = 8)
plot_df = obj@meta.data
plot_df$DiffusionMap_1 = obj@reductions$diffmap@cell.embeddings[,1]
plot_df$DiffusionMap_2 = obj@reductions$diffmap@cell.embeddings[,2]
plot_df_fate = plot_df[!plot_df$celltype %in% c('Spinal cord','Paraxial mesoderm'),]
plot_df_downstream = plot_df[plot_df$celltype %in% c('Spinal cord','Paraxial mesoderm'),]
plot_df_downstream$celltype ='other'
ggplot(plot_df_downstream,aes(x=DiffusionMap_1,y=DiffusionMap_2,color=celltype))+
  geom_point(size=1.2)+scale_color_manual(values = fate_col)+geom_point(data = plot_df_fate)+
  theme_void()+theme(text = element_text(size=16))+theme(legend.position = "none")
ggsave('/data/jiangjunyao/polyATAC/polyatac_figure/all_embedding_no_legend/fig3/NMP diffusion map embedding only fate.pdf',
       width = 10,height = 8)
```

## Plot gene modules
```{r}
col1 <- colorRampPalette(c('#2166AC', '#4393C3','#92C5DE', '#D1E5F0', '#F7F7F7', '#FDDBC7', '#F4A582', '#D6604D', '#B2182B'))(100)
for (i in unique(group_df_20241115$module)) {
  gene_use = group_df_20241115[group_df_20241115$module==i,1]
  mt1 = as.matrix(obj@assays$RNA@data[gene_use,])
  mt1 = colMeans(mt1)
  embed = as.data.frame(obj@reductions$diffmap@cell.embeddings)
  embed$expression = mt1
  ggplot(embed,aes(x=DiffusionMap_1,y=DiffusionMap_2,color=expression))+geom_point(size=1.2)+
    scale_color_gradientn(colors = col1)+theme_void()+
    theme(text = element_text(size=16))+ggtitle(i)+guides(color = "none")
  ggsave(paste0('/data/jiangjunyao/polyATAC/polyatac_figure/single cell fig/NMP_fate_analysis/module_average_expression/diffusionmap embedding_nolegend',i,'.pdf'),
         width = 10,height = 8)
  print(ggplot(embed,aes(x=DiffusionMap_1,y=DiffusionMap_2,color=expression))+geom_point(size=1.2)+
    scale_color_gradientn(colors = col1)+theme_void()+
    theme(text = element_text(size=16))+ggtitle(i))
  ggsave(paste0('/data/jiangjunyao/polyATAC/polyatac_figure/single cell fig/NMP_fate_analysis/module_average_expression/diffusionmap embedding_',i,'.pdf'),
         width = 10,height = 8)
  df_exp = data.frame(c(colMeans(pm_smooth[[2]][gene_use,]),
                        colMeans(sp_smooth[[2]][gene_use,])),
                      c(1:100,1:100),
                      c(rep('Mesoderm lineage',100),rep('Neuron lineage',100)))
  colnames(df_exp) = c('expression','pseudotime','lineage')
  ggplot(df_exp,aes(x=pseudotime,y=expression,color=lineage))+geom_line(size=1)+
    theme_classic()+theme(text = element_text(size=16),plot.title = element_text(hjust = 0.5))+
    scale_color_manual(values = c('#6950a1','#84bf96'))+ggtitle(i)+
    ylab('Scaled expression')+scale_y_continuous(limits = c(0, 2))
  # ggsave(paste0('/data/jiangjunyao/polyATAC/polyatac_figure/single cell fig/NMP_fate_analysis/module_average_expression/lineplot_',i,'.pdf'),
  #        width = 10,height = 8)
}
```

### inter-module level co-expression

```{r}
dual_lineage2 = cbind(pm_smooth[[1]][group_df_20241115[,1],100:1],
                     sp_smooth[[1]][group_df_20241115[,1],])
list_aver = list()
for (i in unique(group_df_20241115$module)){
  gene_use = group_df_20241115[group_df_20241115$module==i,1]
  mt1 = as.matrix(dual_lineage2[gene_use,])
  list_aver[[i]] = colMeans(mt1)
}
module_expression = do.call(bind_cols,list_aver[-1])
result <- rcorr(as.matrix(module_expression))
correlation_matrix <- result$r
adjusted_pvalues <- p.adjust(result$P, method = "fdr")

adjusted_pvalue_matrix <- matrix(adjusted_pvalues, nrow = nrow(correlation_matrix), ncol = ncol(correlation_matrix))
colnames(adjusted_pvalue_matrix) = colnames(correlation_matrix)
rownames(adjusted_pvalue_matrix) = rownames(correlation_matrix)
adjusted_pvalue_matrix[is.na(adjusted_pvalue_matrix)] = 0
col <- colorRampPalette(c('#2166AC', '#4393C3','#92C5DE', '#D1E5F0', '#F7F7F7', '#FDDBC7', '#F4A582', '#D6604D', '#B2182B'))(100)
corrplot(correlation_matrix, 
         method = "circle",
         type = "upper", 
         p.mat = adjusted_pvalue_matrix,
         sig.level = 0.01,insig = "blank",
         pch.cex = 0.8,tl.col = "black",col = col)
print(adjusted_pvalue_matrix)
```

```{r}
legend_colo = list(module=c('Balance regulate'='#78cdd1',
                              'late mesoderm regulate'='#33a3dc',
                              'late neuron regulate'='#77ac98',
                              'early mesoderm regulate'='#90d7ec',
                              "early neuron regulate"='#cde6c7'),
                   celltype=c('Balance NMP'='#4DB6AC',
                              'Paraxial mesoderm biased NMP'='#9ECAE1',
                              'Spinal cord biased NMP'='#C5E1A5',
                              'Paraxial mesoderm'='#0D47A1',
                              'Spinal cord'='#1B5E20'))

col <- colorRampPalette(as.character(jdb_palette('flame_light')))(100)

dual_lineage2 = cbind(pm_smooth[[1]][group_df_20241115$X,100:1],
                     sp_smooth[[1]][group_df_20241115$X,])
no_ba = group_df_20241115[group_df_20241115[,3] != 'Balance regulate',1]
dual_lineage2 = dual_lineage2[no_ba,]
anno_row2 = data.frame(group_df_20241115[group_df_20241115[,3] != 'Balance regulate',3])
colnames(anno_row2) = 'module'
rownames(anno_row2) = rownames(dual_lineage2)
pdf('/data/jiangjunyao/polyATAC/polyatac_figure/single cell fig/NMP_fate_analysis/inter module regulation_single gene_col1.pdf',width = 10,height = 8)
pheatmap(cor(t(dual_lineage2)),cluster_cols = F,cluster_rows = F,
         show_rownames = F,annotation_row =anno_row2,annotation_col = anno_row2,
         annotation_colors = legend_colo,show_colnames = F,color = col)
dev.off()

col <- colorRampPalette(c('#40004B','#762A83','#9970AB','#C2A5CF','#E7D4E8',
                          '#F7F7F7','#D9F0D3','#A6DBA0','#5AAE61','#1B7837','#00441B'))(100)
pdf('/data/jiangjunyao/polyATAC/polyatac_figure/single cell fig/NMP_fate_analysis/inter module regulation_single gene_col2.pdf',width = 10,height = 8)
pheatmap(cor(t(dual_lineage2)),cluster_cols = F,cluster_rows = F,
         show_rownames = F,annotation_row =anno_row2,annotation_col = anno_row2,
         annotation_colors = legend_colo,show_colnames = F,color = col)
dev.off()
```
### Plot single gene expression along trajectory
```{r}
col1 <- colorRampPalette(c('#2166AC', '#4393C3','#92C5DE', '#D1E5F0', '#F7F7F7', '#FDDBC7', '#F4A582', '#D6604D', '#B2182B'))(100)
for (i in group_df_20241115$X ) {
  selected_gene = i
  module_name = group_df_20241115[group_df_20241115$X==i,3]
  mt1 = as.matrix(obj@assays$RNA@data[selected_gene,])
  embed = as.data.frame(obj@reductions$diffmap@cell.embeddings)
  embed$expression = mt1
  ggplot(embed,aes(x=DiffusionMap_1,y=DiffusionMap_2,color=expression))+geom_point(size=1.2)+
    scale_color_gradientn(colors = col1)+theme_void()+
    theme(text = element_text(size=16))+ggtitle(i)+guides(color = "none")
  ggsave(paste0('/data/jiangjunyao/polyATAC/polyatac_figure/single cell fig/NMP_fate_analysis/gene_pseudotime_dynamics/',module_name,'_',
                i,'_UMAP.pdf'),width = 10,height = 8)
  df_exp = data.frame(c(pm_smooth[[2]][selected_gene,],
                        sp_smooth[[2]][selected_gene,]),
                      c(1:100,1:100),
                      c(rep('Mesoderm lineage',100),rep('Neuron lineage',100)))
  colnames(df_exp) = c('expression','pseudotime','lineage')
  ggplot(df_exp,aes(x=pseudotime,y=expression,color=lineage))+geom_line(size=1)+
    theme_classic()+theme(text = element_text(size=16),plot.title = element_text(hjust = 0.5))+
    scale_color_manual(values = c('#6950a1','#84bf96'))+ggtitle(selected_gene)+
    ylab('Scaled expression')+scale_y_continuous(limits = c(0, 2))
  ggsave(paste0('/data/jiangjunyao/polyATAC/polyatac_figure/single cell fig/NMP_fate_analysis/gene_pseudotime_dynamics/',module_name,'_',
                i,'_line.pdf'),width = 10,height = 8)
}
```

### NMP epiblast fate
```{r}
epi_fate <- readRDS("~/polyATAC/all_barcode_kmeans_V2.rds")
nmp_fate <- readRDS("~/polyATAC/nmp_final/only_barcodes_deg/nmp_barcode_fate_kmeans.rds")
rownames(epi_fate) = epi_fate$barcodes
nmp_fate$epi_fate = epi_fate[rownames(nmp_fate),'fate']
df1 = celltype_sample_fraction(nmp_fate,'fate','epi_fate')
df1 = data.frame(c(23.913043,6.521739,7.692308,43.956044),c('PM-NMP','PM-NMP','SC-NMP','SC-NMP'),c('Mesoderm-bias','NeuroEctoderm-bias','Mesoderm-bias','NeuroEctoderm-bias'))
colnames(df1) = c('ratio','NMP_fate','epiblast_fate')
ggplot(df1,aes(x=NMP_fate,y=ratio,fill=epiblast_fate))+geom_col(position = 'dodge',width = 0.6)+theme_minimal()+theme(text = element_text(size=16))+scale_fill_manual(values = c('#AEC7E8FF','#98DF8AFF'))
ggsave('/data/jiangjunyao/polyATAC/polyatac_figure/single cell fig/NMP_fate_analysis/NMP_fate_epiblast_fate_fraction.pdf',width = 8,height = 6)
```

### 